### PR TITLE
fix: truncated actor bundles #3485

### PR DESCRIPTION
**Pull request by aatifsyed**
https://github.com/ChainSafe/forest/issues/3485

### CODE REVIEW

File: src/tool/subcommands/state_migration_cmd.rs [hunk 1]
```diff
@@ -1,140 +1,104 @@
...
-                (b.cid, b.data)
-            })),
+async fn generate_actor_bundle(output: PathBuf) -> anyhow::Result<()> {
+    let (roots, blocks) = stream::iter(ACTOR_BUNDLES.iter())
+        .then(
+            |ActorBundleInfo {
+                 manifest: root,
+                 url,
+             }| async move {
+                info!(%url, "downloading bundle");
```
Comment (aatifsyed) [inline src/tool/subcommands/state_migration_cmd.rs:40]: `forest-tool` should show `tracing` logs

File: src/utils/db/car_stream.rs [hunk 2]
```diff
@@ -26,7 +26,8 @@ pub struct CarHeader {
     pub version: u64,
 }
 
-#[derive(Debug, Clone)]
+// TODO(aatifsyed): too many things are called `block`
```
Comment (aatifsyed) [inline src/utils/db/car_stream.rs:29]:
```suggestion
```

Too many things are called `Block` in `forest`

File: src/networks/mod.rs [hunk 3]
```diff
@@ -360,47 +360,54 @@ pub struct ActorBundleInfo {
...
+    fn unwrap_parse(manifest: &str, url: &str) -> Self {
+        Self {
+            manifest: manifest.parse().unwrap(),
+            url: url.parse().unwrap(),
+        }
+    }
+}
+
+pub static ACTOR_BUNDLES: Lazy<Box<[ActorBundleInfo]>> = Lazy::new(|| {
+    Box::new([
```
Comment (elmattic) [inline src/networks/mod.rs:373]: All the cids are the same right? Maybe keep the comments (devnet, calibnet, etc) for clarity.

**Comment by aatifsyed** [comment]
> So the new actor bundle file is `12139433` bytes big? It's almost six time the previous size.

Not surprising - the uncompressed is 60MiB:
```python
#!/usr/bin/env python3

import requests
import humanize

urls = [
    "https://github.com/filecoin-project/builtin-actors/releases/download/v9.0.3/builtin-actors-calibrationnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v10.0.0-rc.1/builtin-actors-calibrationnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v11.0.0-rc2/builtin-actors-calibrationnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v9.0.3/builtin-actors-devnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v10.0.0/builtin-actors-devnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v11.0.0/builtin-actors-devnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v9.0.3/builtin-actors-mainnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v10.0.0/builtin-actors-mainnet.car",
    "https://github.com/filecoin-project/builtin-actors/releases/download/v11.0.0/builtin-actors-mainnet.car",
]

total = 0

for url in urls:
    response = requests.head(url, allow_redirects=True)
    response.raise_for_status()
    content_length = response.headers["content-length"]
    total += int(content_length)

print(humanize.naturalsize(total, binary=True))
```
```
60.4 MiB
```

File: src/tool/subcommands/state_migration_cmd.rs [hunk 4]
```diff
@@ -1,140 +1,104 @@
...
-        writer.close().await?;
-    }
-    if is_bundle_valid(root, BufReader::new(async_fs::File::open(&tmp).await?)).await? {
-        tmp.persist(&cached_car_path)?;
-        Ok(cached_car_path)
-    } else {
-        anyhow::bail!("Invalid bundle: {url}");
-    }
-}
+    let car_zst = zstd::encode_all(car.as_slice(), 3)?;
```
Comment (lemmih) [inline src/tool/subcommands/state_migration_cmd.rs:86]: This should reduce the file size by ~90%.

```suggestion
    let car_zst = zstd::encode_all(car.as_slice(), 17)?;
```

File: src/tool/subcommands/state_migration_cmd.rs [hunk 5]
```diff
@@ -1,140 +1,104 @@
...
-                return Ok(cached_car_path);
-            }
-        }
+        .try_collect::<Vec<_>>()
+        .await?
+        .into_iter()
+        .unzip::<_, _, Vec<&Cid>, Vec<_>>();
+
+    let roots = roots.into_iter().cloned().collect::<CidHashSet>();
+    let blocks = blocks.into_iter().flatten().try_fold(
```
Comment (lemmih) [inline src/tool/subcommands/state_migration_cmd.rs:60]: As per our discussion, let's sort the blocks such that the bundle is deterministic.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Review by ruseinov — APPROVED** [review]
Looks good and way more readable. 
Would be nice to have some details/comments about the car validation though.
it's not really clean why certain assumptions are enforced.

File: src/ipld/cid_hashset.rs [hunk 6]
```diff
@@ -1,16 +1,17 @@
...
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use crate::ipld::CidHashMap;
+use crate::utils::cid::SmallCid;
+use ahash::HashSet;
 use cid::Cid;
 
 #[derive(Default, Clone)]
-pub struct CidHashSet(CidHashMap<()>);
+pub struct CidHashSet(HashSet<SmallCid>);
```
Comment (lemmih) [inline src/ipld/cid_hashset.rs:9]: I am not sold on this. It wastes a lot of memory. Let's discuss this in slack.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: build.rs [hunk 7]
```diff
@@ -10,44 +9,9 @@ const PROTO_DIR: &str = "proto";
 const CARGO_OUT_DIR: &str = "proto";
 
 fn main() -> anyhow::Result<()> {
-    ensure_actor_bundle_includable()?;
     generate_protobuf_code()
 }
 
-fn ensure_actor_bundle_includable() -> anyhow::Result<()> {
```
Comment (LesnyRumcajs) [inline build.rs:17]: Why don't we need this check anymore?

File: src/ipld/cid_hashset.rs (hunk 6 above)
Comment (aatifsyed) [reply src/ipld/cid_hashset.rs:9]: https://github.com/ChainSafe/forest/issues/3498

File: build.rs (hunk 7 above)
Comment (aatifsyed) [reply build.rs:17]: It's covered by the `asset_integrity` test now

File: src/ipld/cid_hashset.rs (hunk 6 above)
Comment (lemmih) [reply src/ipld/cid_hashset.rs:9]: I don't feel this is resolved.

Comment (aatifsyed) [reply src/ipld/cid_hashset.rs:9]: Blocked on #3502

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/tool/subcommands/state_migration_cmd.rs [hunk 8]
```diff
@@ -2,129 +2,102 @@
...
+    ensure!(roots.iter().all_unique());
 
-    let all_roots = car_streams
-        .iter()
-        .flat_map(|it| it.header.roots.iter())
-        .unique()
-        .cloned()
-        .collect::<Vec<_>>();
+    roots.sort(); // deterministic
+    let blocks = blocks.into_iter().flatten().try_fold(
```
[the remark this answers is not in the document]
Comment (aatifsyed) [reply src/tool/subcommands/state_migration_cmd.rs:61]: `blocks` is a `Vec<Vec<CarBlock>>` - sorting and deduping wouldn't give us the behaviour I want.

I want a strict (no clobbering) collection into `BTreeMap<Cid, Vec<u8>>`

**Review by lemmih — APPROVED** [review]
Great!

**Pull request merged**